extraConfigDpuInfra.py

```diff
@@ -118,9 +118,10 @@ def run_dpu_network_operator_git(lh: host.Host, kc: str) -> None:
     cur_dir = os.getcwd()
     os.chdir(repo_dir)
     lh.run("rm -rf bin")
-    env = os.environ.copy()
-    env["KUBECONFIG"] = kc
-    env["IMG"] = "quay.io/wizhao/dpu-network-operator:Nov1_WZ_DPU_DS_Test_1"
+    env = {
+        "KUBECONFIG": kc,
+        "IMG": "quay.io/wizhao/dpu-network-operator:Nov1_WZ_DPU_DS_Test_1",
+    }
```

Comment:
Will this work? We are missing the rest of the env

Reply:
commit message name:  Change semantics of env in Host.run.

New question: How can I remove stuff from the env on remote host with the new semantics? Suppose I want to remove everything and only have 1 variable.

Reply:
quote commit message:

>     Also, "env" accepts setting a dictionary value to None, which means to
>    unset the environment variable.

Reply:
Suppose the env contains keys A B C, and I want it only to contain D. How do I do that?

Reply:
As you care about the keys `"A"`, `"B"`, `"C"`, `"D"`, you do
```python
    env = {
        "A": None,
        "B": None,
        "C": None,
        "D": "d-value",
    }
```

As said, replacing the entire environment as before does not make sense (and you only care about the keys  `"A"`, `"B"`, `"C"`, `"D"` you know about). Suppose you really really want to simulate the old behavior, then:
```python
     env = {
        "D": "d-value",
     }
     for k in os.environ:
         if k not in env:
             env[k] = None
```
That then only works with `Host._run_local()`. It won't do anything useful when running with `Host.need_sudo()` or as ssh remote. But that is also true about the current `main` branch...

Reply:
if i pass in an `env`, doesn't it make sense to have it be the full thing I ask to be run on the other side (ssh or root)?

Reply:
I think it does not.

sudo and ssh will prepare environment variables that are suitable in that context. E.g. `USER`, `PWD`, `PATH`. It makes not sense trying to delete those (or replace them with the value you get locally from `os.environ`).

Of course, if you explicitly set `env = { "PATH": value, "USER": "me" }`, then they will be set as you'd expect. But what is the use of that?

Set/Unset the variables that you know you want to set, and leave the others alone. Especially in sudo/ssh, you already run in a sanitized environment that has only variables that are supposed to be there. Leave those alone.

If you run locally, and you want to sanitze the environment, then dropping *all* unknown variables may make some sense. The previous comment shows how that can be done. Note that no callers exist that would want to do that.

Reply:
My expectation is this:

If I pass in {} as an env, I expect the env to be empty
If I pass in {"a":"b"} as an env, I expect the env to be exactly that
If I don't pass in an env, the env should be whatever it is.

Reply:
How would you clear the env? i.e. I don't know what's in the env but I want to remove everything.

Reply:
> If I pass in {} as an env, I expect the env to be empty
> If I pass in {"a":"b"} as an env, I expect the env to be exactly that


> How would you clear the env? i.e. I don't know what's in the env but I want to remove everything.

In my last comments I said how you can do it. But the real answer is: you don't. Why do you want to do that?

For example, [paramiko.Client's `exec_command()`](https://docs.paramiko.org/en/latest/api/client.html) also has an "environment" argument, which behaves like `env` on this branch.

Reply:
I don't follow how your last command explains how to remove keys you don't know exist.

Reply:
If the server is setting some var, this will not unset it.

Reply:
> If the server is setting some var, this will not unset it.

I know. In the same comment I said:

> That then only works with Host._run_local(). It won't do anything useful when running with Host.need_sudo() or as ssh remote. But that is also true about the current main branch...

And in [this comment](https://github.com/bn222/cluster-deployment-automation/pull/163#discussion_r1567148761), why that is not useful to do.

Don't do that. Why do you want to do that?

Reply:
Because it's a side effect. That's why it's better to pass in the env as a whole, not as a diff. For that reason, we have code that first copies the env and adds to it if that's what you want.

Reply:
The environment that is setup for you via `sudo`/`ssh` is clean and prepared to be suitable. That should not be replaced or cleaned up. In any case, with *what* would you want to replace it?


Btw, *all* callers of `Host.run()` currently take the existing environment and modify it (thereby keeping all unknown keys). None of the callers handles this "side-effect" by preparing a clean run environment in the local case (they get the caller's environment). On the other hand, with `sudo`/`ssh` you get a clean, well defined environment already.